flight/Modules/VTXConfig/VTXConfig.c

```diff
@@ -0,0 +1,317 @@
+	}
+
+	return 0;
+}
+
+MODULE_INITCALL(VTXConfigInitialize, VTXConfigStart);
+
+
+/**
+ * Main VTXConfig Task
+ */
+
+static void vtxConfigTask(void *parameters)
+{
+	VTXInfoData info;
+	VTXSettingsData settings;
+
+	uint16_t baud_rate = TBS_MIN_BAUD;
+	uint8_t n_fails = 0;
+	enum STATE state = DISCONNECTED;
+	enum STATE state_prev = DISCONNECTED;
+	bool update_vtx = false;
+
+	VTXSettingsGet(&settings);
+
+	// Wait for power to stabilize before talking to external devices
+	PIOS_Thread_Sleep(1000);
+
+	// Loop forever
+	while (1) {
+		// Try to connect faster when disconnected
+		if (state == CONNECTED) {
+			PIOS_Thread_Sleep(1000);
+		}
+		else {
+			PIOS_Thread_Sleep(500);
+		}
+
+		if (settings_updated) {
+			VTXSettingsGet(&settings);
+			settings_updated = false;
+			update_vtx = true;
+		}
+
+		if (tbsvtx_get_state(vtxConfigPort, &info) >= 0) {
+			state = CONNECTED;
+			n_fails = 0;
+			info.State = VTXINFO_STATE_CONNECTED;
+			VTXInfoSet(&info);
+			if (state_prev == DISCONNECTED) {
+				update_vtx = true;
+			}
+		}
+		else {
+			n_fails += 1;
+		}
+
+		if (n_fails > MAX_FAILS) {
+			state = DISCONNECTED;
+			memset((void*)&info, 0, sizeof(info));
+			info.State = VTXINFO_STATE_DISCONNECTED;
+			VTXInfoSet(&info);
+		}
+
+		if (state == CONNECTED) {
+			PIOS_Thread_Sleep(500);
+			//update_vtx = true;
+			if ((settings.Mode == VTXSETTINGS_MODE_ACTIVE) && update_vtx) {
+
+				// Get the frequency
+				uint16_t frequency;
+				switch (settings.Band) {
+					case VTXSETTINGS_BAND_BAND5G8A:
+						frequency = BAND_5G8_A_FREQS[settings.Band_5G8_A_Frequency];
+						break;
+					case VTXSETTINGS_BAND_BAND5G8B:
+						frequency = BAND_5G8_B_FREQS[settings.Band_5G8_B_Frequency];
+						break;
+					case VTXSETTINGS_BAND_BAND5G8E:
+						frequency = BAND_5G8_E_FREQS[settings.Band_5G8_E_Frequency];
+						break;
+					case VTXSETTINGS_BAND_AIRWAVE:
+						frequency = AIRWAVE_FREQS[settings.Airwave_Frequency];
+						break;
+					case VTXSETTINGS_BAND_RACEBAND:
+						frequency = RACEBAND_FREQS[settings.Raceband_Frequency];
+						break;
+					default:
+						update_vtx = false;
+				}
+				if (update_vtx) {
+					update_vtx = false;
+					if (frequency != info.Frequency) {
+						if (tbsvtx_set_freq(vtxConfigPort, frequency) < 0) {
+							// something went wrong, we need to re-try again
+							update_vtx = true;
+						}
+						PIOS_Thread_Sleep(500);
+					}
+
+					// Don't increase the power if the VTX reports 0mW. This is a special mode
+					// e.g. "pit mode" for TBS and the user requested to transmit at low power.
+					if ((info.Power != 0) && (VTX_POWER[settings.Power] != info.Power)) {
+						if (tbsvtx_set_power(vtxConfigPort, VTX_POWER[settings.Power]) < 0) {
+							// somethign went wrong, we need to re-try again
+							update_vtx = true;
+						}
+					}
+				}
+			}
+		}
+		else {
+			// Cycle through different baud rates. The spec says 4.8kbps, but it can deviate
+			if (baud_rate < TBS_MAX_BAUD) {
```

Comment:
For sure. I figured it would be faster / more reliable to get it working with an UART first but we should look into using a software based UART using the GPIO pin. The baudrate is very low, so it should be feasible.